src/meta/src/barrier/checkpoint/recovery.rs

```diff
@@ -0,0 +1,406 @@
+                DatabaseRecoveringStage::Resetting { .. } => state,
+            },
+        };
+        let DatabaseRuntimeInfoSnapshot {
+            database_fragment_info,
+            mut state_table_committed_epochs,
+            subscription_info,
+            mut stream_actors,
+            mut source_splits,
+            mut background_jobs,
+        } = runtime_info;
+        let result: MetaResult<_> = try {
+            control_stream_manager.add_partial_graph(self.database_id, None)?;
+            control_stream_manager.inject_database_initial_barrier(
+                self.database_id,
+                database_fragment_info,
+                &mut state_table_committed_epochs,
+                &mut stream_actors,
+                &mut source_splits,
+                &mut background_jobs,
+                subscription_info,
+                None,
+                &self.control.hummock_version_stats,
+            )?
+        };
+        match result {
+            Ok((remaining_workers, database, prev_epoch)) => {
+                status.stage = DatabaseRecoveringStage::Initializing {
+                    remaining_workers,
+                    database,
+                    prev_epoch,
+                };
+            }
+            Err(e) => {
+                warn!(database_id = self.database_id.database_id,e = ?e.as_report(), "failed to inject initial barrier");
+                let (backoff_future, reset_request_id) = status.next_retry();
+                let remaining_workers =
+                    control_stream_manager.reset_database(self.database_id, reset_request_id);
+                status.stage = DatabaseRecoveringStage::Resetting {
+                    remaining_workers,
+                    reset_workers: Default::default(),
+                    reset_request_id,
+                    backoff_future: Some(backoff_future),
+                };
+            }
+        }
+    }
+
+    pub(crate) fn remove(self) {
+        self.control
+            .databases
+            .remove(&self.database_id)
+            .expect("should exist");
+    }
+}
+
+pub(crate) struct EnterRunning;
+
+impl DatabaseStatusAction<'_, EnterRunning> {
+    pub(crate) fn enter(self) {
+        let database_status = self
+            .control
+            .databases
+            .get_mut(&self.database_id)
+            .expect("should exist");
+        match database_status {
+            DatabaseCheckpointControlStatus::Running(_) => {
+                unreachable!("should not enter running again")
+            }
+            DatabaseCheckpointControlStatus::Recovering(state) => {
+                let temp_place_holder = DatabaseRecoveringStage::Resetting {
+                    remaining_workers: Default::default(),
+                    reset_workers: Default::default(),
+                    reset_request_id: 0,
+                    backoff_future: None,
+                };
+                match replace(&mut state.stage, temp_place_holder) {
```

Comment:
nits: What do we need the `replace` here? The whole `database_status` will be reset in L400 anyway so I guess we only need to do `match state.stage` here?